synapse/handlers/device.py

```diff
@@ -595,7 +606,7 @@ async def rehydrate_device(
 
 
 def _update_device_from_client_ips(
-    device: JsonDict, client_ips: Dict[Tuple[str, str], JsonDict]
+    device: JsonDict, client_ips: Mapping[Tuple[str, str], Mapping[str, Any]]
```

Comment:
`client_ips` comes from the return of `get_last_client_ip_by_device`. I downgraded the `Dict`s to `Mapping`s to make mypy happy.

Reply:
Why was mypy unhappy here, sorry---something to do with a `Dict` being mutable?

Reply:
It's to do with the inner `JsonDict` not being the same as `LastClientIpByDeviceEntry`:
`Argument 2 ... has incompatible type "Dict[Tuple[str, str], LastClientIpByDeviceEntry]"; expected "Dict[Tuple[str, str], Dict[str, Any]]"  [arg-type]`